File: Striver/Trees/Q12.py
# diameter of a binary tree

from Node import Node

def height(root, ans):
    if not root:
        return 0
    lh = height(root.left, ans)
    rh = height(root.right, ans)
    ans[0] = max(ans[0], 1 + lh + rh)
    return 1 + max(lh, rh)

def diameter(root):
    if not root:
        return 0
    ans = [float('-inf')]
    height(root, ans)
    return ans[0]

root = Node(1)
root.left = Node(2)
root.right = Node(3)
root.left.left = Node(4)
root.left.left.left = Node(8)
root.left.left.right = Node(9)
root.left.right = Node(5)
root.right.left = Node(6)
root.right.right = Node(7)

print(diameter(root))

File: Striver/Trees/Q12.java
package Striver.Trees;

public class Q12{
    // Diameter of a Binary Tree

    public int diameterOfBinaryTree(Node root) {
        if(root == null) return 0;
        int[] ans = new int[1];
        height(root, ans);
        return ans[0];
    }

    public int height(Node root, int[] ans){
        if(root == null) return 0;
        int left = height(root.left, ans);
        int right = height(root.right, ans);
        ans[0] = Math.max(ans[0], left + right);
        return Math.max(left, right) + 1;
    }

    public static void main(String[] args) {
        Q12 sol = new Q12();
        Node root = new Node(1);
        root.left = new Node(2);
        root.left.left = new Node(4);
        root.right = new Node(3);
        System.out.println(sol.diameterOfBinaryTree(root));   
    }
}